sdk/src/main/java/com/mapbox/maps/ViewAnnotationManagerImpl.kt

```diff
@@ -178,13 +195,94 @@ internal class ViewAnnotationManagerImpl(
 
     if (viewAnnotationOptions.isEmpty()) return null
     val coordinates = coordinatesFromAnnotations(viewAnnotationOptions)
-    val paddings = calculateEdgeInsets(viewAnnotationOptions, edgeInsets)
-    return mapboxMap.cameraForCoordinates(
+    var cameraOptionForCoordinates = mapboxMap.cameraForCoordinates(
       coordinates,
-      paddings,
+      EdgeInsets(0.0, 0.0, 0.0, 0.0),
       bearing,
       pitch
     )
+
+    var isCorrectBound = false
+    // keep north, east, west and southmost Annotation and its frame to adjust paddings for zoom.
+    var north: Pair<ViewAnnotationOptions, Rect?>? = null
+    var east: Pair<ViewAnnotationOptions, Rect?>? = null
+    var west: Pair<ViewAnnotationOptions, Rect?>? = null
+    var south: Pair<ViewAnnotationOptions, Rect?>? = null
+
+    while (!isCorrectBound) {
```

Comment:
how many iteration can be here in worst case? can it be forever loop?

Reply:
Can't be a forever loop, the worst case iteration is 2 loops. 
Initially, we are getting innerBounds `cameraOption` before while loop, then the next iteration is checking if it needs extra paddings. All other nullability checks have already been tested.

Reply:
We discovered a case where this can loop forever on iOS - with certain combination of camera state/map size/view annotations there might be one pair(view annotation options & rect) to be e.g. southern-most for a particular zoom level, and for another zoom level the different pair can become southern-most.

Potentially this code is also susceptible to the same scenario, worthwhile to double check. @maios is leading the investigation of this on iOS.

Reply:
I'd probably removed the loop, extracted the code calculating `cameraOptionForCoordinates` and explicitly called it twice to avoid any risks.

Reply:
Yeah, make sense. although I couldn't reproduce the issue on android with the same setup. I will try to loop it with a counter and that would be our maximum tries to adjust the bounds. 
I can see that this might not provide correct bounds for all cases, but we have discussed this with gl-native to improve this in next releases.

Reply:
added [ticket](https://mapbox.atlassian.net/browse/MAPSAND-649) to track this issue.

Reply:
@ank27 at the same time why do we need a loop for two iterations? What about extracting to function and calling twice? Code will be more clear and not additional variables involved

Reply:
In case we need to have more iteration in the future to have better bounds. but that's debatable. 
btw why can't we use loop though?

Reply:
@ank27 I was thinking that the loop is not needed if we know that we have fixed and limited number (== 2) of iterations to calculate the bounds. Now it seems that it's not - and I'm actually ok with having the loop with counter which can be controlled. The question is do we want it to be 2 or some larger value to be almost certain that the bounds will be correctly calculated.

Reply:
2 iteration i found good for the cases i have tested, but i can't be sure of all edge-cases (for eg. pitched map and with bearing). so i am not 100% sure about fixed number. 

> The question is do we want it to be 2 or some larger value to be almost certain that the bounds will be correctly calculated.

cut a [ticket](https://mapbox.atlassian.net/browse/MAPSAND-649) to look more into finding correct calculations for all cases.